x/leverage/keeper/msg_server.go

```diff
@@ -83,6 +83,44 @@ func (s msgServer) Withdraw(
 	}, err
 }
 
+func (s msgServer) WithdrawMaximum(
+	goCtx context.Context,
+	msg *types.MsgWithdrawMaximum,
+) (*types.MsgWithdrawMaximumResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	supplierAddr, err := sdk.AccAddressFromBech32(msg.Supplier)
+	if err != nil {
+		return nil, err
+	}
+
+	uToken, err := s.keeper.maxWithdraw(ctx, supplierAddr, msg.Denom)
+	if err != nil {
+		return nil, err
+	}
+
+	received, err := s.keeper.Withdraw(ctx, supplierAddr, uToken)
+	if err != nil {
+		return nil, err
+	}
+
+	s.keeper.Logger(ctx).Debug(
+		"maximum supplied assets withdrawn",
+		"supplier", msg.Supplier,
+		"redeemed", uToken.String(),
+		"received", received.String(),
+	)
+	err = ctx.EventManager().EmitTypedEvent(&types.EventWithdraw{
+		Supplier: msg.Supplier,
+		Utoken:   uToken,
+		Asset:    received,
+	})
```

Comment:
lines 102-117 could go to a new common function (you just need to add debug log msg argument to that function).

Reply:
done